cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -806,6 +806,45 @@ class ListConverter
   bool strict_conversions_;
 };
 
+// ----------------------------------------------------------------------
+// Convert maps
+
+// Define a MapConverter as a ListConverter that uses MapBuilder.value_builder
+// to append struct of key/value pairs
+template <NullCoding null_coding>
+class MapConverter : public ListConverter<MapType, null_coding> {
+ public:
+  using BASE = ListConverter<MapType, null_coding>;
+
+  explicit MapConverter(bool from_pandas, bool strict_conversions)
+      : BASE(from_pandas, strict_conversions) {}
+
+  Status AppendSingleVirtual(PyObject* obj) override {
+    RETURN_NOT_OK(BASE::AppendSingleVirtual(obj));
+    return VerifyLastStructAppended();
+  }
+
+  Status AppendMultiple(PyObject* seq, int64_t size) override {
+    RETURN_NOT_OK(BASE::AppendMultiple(seq, size));
+    return VerifyLastStructAppended();
+  }
+
+  Status AppendMultipleMasked(PyObject* seq, PyObject* mask, int64_t size) override {
+    RETURN_NOT_OK(BASE::AppendMultipleMasked(seq, mask, size));
+    return VerifyLastStructAppended();
+  }
+
+ protected:
+  Status VerifyLastStructAppended() {
```

Comment:
Is it useful to do this check at every item? Why not once at the end?

Reply:
That was my first thought, but the struct data ends up wrong in the list. I think it's because the struct builder length gets used for the list offsets, so it needs to be called at the end of appending items to each map "slot."

Reply:
I don't get it. The following works currently:
```python
>>> ty = pa.list_(pa.struct([('key', pa.string()), ('item', pa.int8())]))                                                                                                
>>> arr = pa.array([None, [], [('foo', 1), ('bar', 2)], [('xyzzy', 3)]], type=ty)                                                                                        
>>> arr                                                                                                                                                                  
<pyarrow.lib.ListArray object at 0x7fc1bfe962e8>
[
  null,
  -- is_valid: all not null
  -- child 0 type: string
    []
  -- child 1 type: int8
    [],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "foo",
      "bar"
    ]
  -- child 1 type: int8
    [
      1,
      2
    ],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "xyzzy"
    ]
  -- child 1 type: int8
    [
      3
    ]
]
```

Map type conversion is morally the same, so it doesn't seem right that it needs such a special treatment.

Reply:
When building a Map array from python, the above works fine with the MapBuilder too. The problem is in C++, when the MapBuilder `key_builder()` and `item_builder()` are used. I will add a C++ test that will hopefully make it clearer.

Reply:
But this method is only doing a check, not changing everything. Why is it required to call it at every element?

Reply:
It doesn't just do a check, it adjust the struct_builder length, which is used when the `ListBuilder.Append()` is called (`value_builder` below would be the `struct_builder`):

```C++
Status AppendNextOffset() {
    ARROW_RETURN_NOT_OK(CheckNextOffset());
    const int64_t num_values = value_builder_->length();
    return offsets_builder_.Append(static_cast<offset_type>(num_values));
  }
```

It needs to be called every element in the Map so the offsets of the underlying list are correct. If it's only called once at the end, then the list offsets would say all values are in the last element. Doing this for the test `BuildingStringToInt` would fail with this:


```C++
std::vector<int32_t> offsets = {0, 2, 2, 3, 3};
auto expected_keys = ArrayFromJSON(utf8(), R"(["joe", "mark", "cap"])");
auto expected_values = ArrayFromJSON(int32(), "[0, null, 8]");
```
```diff
/home/bryan/git/arrow/cpp/src/arrow/testing/gtest_util.cc:84: Failure
Failed
@@ -0, +0 @@
-[]
+[{key: "joe", value: 0}, {key: "mark", }]
@@ -2, +2 @@
+[{key: "cap", value: 8}]
@@ -3, +4 @@
-[{key: "joe", value: 0}, {key: "mark", }, {key: "cap", value: 8}]
```
So all the list values end up at the last element because the struct_builder length was 0 when building the offsets at each element.

Reply:
Hmm... can you please explain me where in the following code it does the adjustement?
```c++

    auto struct_builder = checked_cast<StructBuilder*>(BASE::value_converter_->builder());
    auto key_builder = struct_builder->field_builder(0);
    if (key_builder->null_count() > 0) {
      return Status::Invalid("Invalid Map: key field can not contain null values");
    }
    return Status::OK();
```

Reply:
omg, sorry. I misunderstood from the get-go and I thought this was a question for another part in the code, let me start again.

Yes, you're right this just checks that the map key value is never null. It could be done once at the end, but I thought it would be better to fail fast rather than process the entire sequence before checking that there is a null key.

The difference between Map and List of Structs is that for Map you could not have the following, which this code would check for:

```python
In [7]: arr = pa.array([None, [(None, 2)], [('foo', 1)], []], type=ty)                                                       

In [8]: arr                                                                                                                  
Out[8]: 
<pyarrow.lib.ListArray object at 0x7fdea79a2e88>
[
  null,
  -- is_valid: all not null
  -- child 0 type: string
    [
      null
    ]
  -- child 1 type: int8
    [
      2
    ],
  -- is_valid: all not null
  -- child 0 type: string
    [
      "foo"
    ]
  -- child 1 type: int8
    [
      1
    ],
  -- is_valid: all not null
  -- child 0 type: string
    []
  -- child 1 type: int8
    []
]
```

Reply:
Well, perhaps at least cache the `key_builder` so that you aren't looking it up every time?

Reply:
yeah, good idea. will do